### ISSUE TITLE
eth_call doesn't work with geth 1.13

### ISSUE THREAD

**Issue by nuliknol** [NomicFoundation/hardhat#4648]
### Version of Hardhat

2.19.1

### What happened?

When sending eth_call to a contract for accessing read-only function with golang of Ethereum geth 1.13, the response from Hardhat is "unknown selector" error. 

The reason for this error is that Hardhat expect contract input provided in the field `data` , but ethereum geth 1.13 is sending it in field name `input`. 

This is RPC message sent with golang of geth 1.13.5

```
{"jsonrpc":"2.0","id":1,"method":"eth_call","params":[
    {   
        "from":"0x0000000000000000000000000000000000000000",
        "input":"0x8b1329e0",
        "to":"0x5fbdb2315678afecb367f032d93f642f64180aa3"
    },  
    "latest"
]}
```

and this is RPC message sent with hardhat `run` command:

```
{"jsonrpc":"2.0","method":"eth_call","params":[
    {   
        "from":"0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266",
        "to":"0x5fbdb2315678afecb367f032d93f642f64180aa3",
        "data":"0x8b1329e0"
    },  
    "latest"
],"id":4}
```

As you can see Hardhat encodes the RPC data with `data` field, while geth uses `input` field. For this reason I am getting `unknown selector` error because Hardhat simply doesn't read contract input when decoding JSON text sent over HTTP

How to reproduce: 
1. Create a contract with read-only function
2. Download geth 1.13 and generate contract wrappers using `abigen` tool
3. Create a program in golang to send eth_call to the read-only function

The result is "unknow selector" error and execution of a fallback function

### Minimal reproduction steps

Hardhat expects contract input in field named `data` while Geth 1.12 or 1.13 uses field name `input` for sending contract's input.

this is RPC message sent with hardhat using `run` command:
```
{"jsonrpc":"2.0","method":"eth_call","params":[
    {   
        "from":"0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266",
        "to":"0x5fbdb2315678afecb367f032d93f642f64180aa3",
        "data":"0x8b1329e0"
    },  
    "latest"
],"id":4}
```

and this is RPC message sent with golang of geth 1.12 or 1.13 ;

```
{"jsonrpc":"2.0","id":1,"method":"eth_call","params":[
    {   
        "from":"0x0000000000000000000000000000000000000000",
        "input":"0x8b1329e0",
        "to":"0x5fbdb2315678afecb367f032d93f642f64180aa3"
    },  
    "latest"
]}
```
As you can see Hardhat uses `data` field name, while the correct is to use `input` field name

here is how transaction encoding json format is defined in geth:
```
type txJSON struct {
    Type hexutil.Uint64 `json:"type"`

    ChainID              *hexutil.Big    `json:"chainId,omitempty"`
    Nonce                *hexutil.Uint64 `json:"nonce"`
    To                   *common.Address `json:"to"`
    Gas                  *hexutil.Uint64 `json:"gas"`
    GasPrice             *hexutil.Big    `json:"gasPrice"`
    MaxPriorityFeePerGas *hexutil.Big    `json:"maxPriorityFeePerGas"`
    MaxFeePerGas         *hexutil.Big    `json:"maxFeePerGas"`
    MaxFeePerBlobGas     *hexutil.Big    `json:"maxFeePerBlobGas,omitempty"`
    Value                *hexutil.Big    `json:"value"`
    Input                *hexutil.Bytes  `json:"input"`
    AccessList           *AccessList     `json:"accessList,omitempty"`
    BlobVersionedHashes  []common.Hash   `json:"blobVersionedHashes,omitempty"`
    V                    *hexutil.Big    `json:"v"`
    R                    *hexutil.Big    `json:"r"`
    S                    *hexutil.Big    `json:"s"`
    YParity              *hexutil.Uint64 `json:"yParity,omitempty"`

    // Only used for encoding:
    Hash common.Hash `json:"hash"`
}
```
For this reason when using geth 1.12 or 1.13 the method calls to read-only function will fail with "unknown selector" error, because the field `input` is not being read by hardhat

**Comment by marlon-wiprud** [comment]
Ran into a similar issue making read-only calls against a hardhat contract with go bindings via abigen - downgrading from `1.13.5` to `1.12.2` resolved the problem for me.

**Comment by nuliknol** [comment]
> Ran into a similar issue making read-only calls against a hardhat contract with go bindings via abigen - downgrading from `1.13.5` to `1.12.2` resolved the problem for me.
> 
> edit: found this PR which has aims to fix the problem: #4438

yeah, the 1.12 works, but it shouldn't because the change in core/types of was made in 1.12 , but another thing is that the field was always was called "input" since 1.10 . I think the encoding in the RPC message is not the issue, the issue must be somewhere else.  luckily it works for 1.12 at least

**Comment by marlon-wiprud** [comment]
@nuliknol another problem I'm running into for what its worth - I'm able to make contract calls with geth 1.12.2 -> hardhat 2.19 but not receiving any events via websocket subscriptions to the same contract.

Changing the geth version to 1.13.5 reverses the problem - failure to make contract calls, but can receive websocket events
